File: SnakeModel/SnakeGame.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace SnakeModel
{
    public class SnakeGame
    {

        public const int SIZE = 20;
        public Snek Snake { get; private set; }
        public int Score { get { return Snake.Pieces.Count; } }

        public int FoodX { get; private set; }
        public int FoodY { get; private set; }

        public bool GameOver { get; private set; }

        private Random random;

        public SnakeGame()
        {
            random = new Random();
            Snake = new Snek(random.Next(SIZE - 4) + 4, random.Next(SIZE - 4) + 4, 3, Directions.UP);
            SpawnFood();
        }

        public void Update()
        {
            if (GameOver) return;
            Snake.Update();
            CheckForFood();
            CheckBodyCollisions();
        }

        public void Goto(int x, int y)
        {
            var headDir = Snake.Head.Direction;
            var head = Snake.Head;

            if(headDir.Y == 0)
            {
                if (y > head.Y)
                    head.Direction = new Direction(Directions.DOWN);
                else
                    head.Direction = new Direction(Directions.UP);
            }
            else
            {
                if (x > head.X)
                    head.Direction = new Direction(Directions.RIGHT);
                else
                    head.Direction = new Direction(Directions.LEFT);
            }

            head.NextDirection = head.Direction;
        }

        private void CheckForFood()
        {
            if(Snake.Head.X == FoodX && Snake.Head.Y == FoodY)
            {
                SpawnFood();
                Snake.Eat();
            }
        }

        private void CheckBodyCollisions()
        {
            GameOver = Snake.Pieces.Exists(x => x.X == Snake.Head.X && x.Y == Snake.Head.Y);
        }

        private void SpawnFood()
        {
            bool valid = true;
            do
            {
                FoodX = random.Next(SIZE - 1);
                FoodY = random.Next(SIZE - 1);

                if (Snake.Head.X == FoodX && Snake.Head.Y == FoodY) valid = false;

                foreach (var piece in Snake.Pieces)
                    if (piece.X == FoodX && Snake.Head.Y == FoodY) valid = false;

            } while (!valid);
        }
    }
}

File: android/snakewear/wearsnakecommon/src/main/java/se/nocroft/wearsnakecommon/model/SnakeGame.java
package se.nocroft.wearsnakecommon.model;

import java.util.Random;

public class SnakeGame {

    public final static int SIZE = 20;

    private Snek snake;

    private int foodX;
    private int foodY;

    public boolean gameOver = false;

    private Random random;

    public SnakeGame() {
        random = new Random();
        snake = new Snek(
                random.nextInt(SIZE - 8) + 8,
                random.nextInt(SIZE - 8) + 8, 3, Direction.Directions.UP);
        spawnFood();
    }

    public void update() {
        if (gameOver) return;
        snake.update();
        checkBodyCollisions();
        if (gameOver) return;
        checkForFood();
    }

    public void changeDirection(int x, int y) {
        Direction headDir = snake.getHead().getDirection();
        SnekPiece head = snake.getHead();

        if (headDir.getY() == 0) {
            if (y > head.getY())
                head.setDirection(new Direction(Direction.Directions.DOWN));
            else
                head.setDirection(new Direction(Direction.Directions.UP));
        } else {
            if (x > head.getX())
                head.setDirection(new Direction(Direction.Directions.RIGHT));
            else
                head.setDirection(new Direction(Direction.Directions.LEFT));
        }

        head.setNextDirection(head.getDirection());
    }

    public Snek getSnake() {
        return snake;
    }

    public int getScore() {
        return snake.getBody().size() + 1;
    }

    public boolean isGameOver() {
        return gameOver;
    }

    public int getFoodX() {
        return foodX;
    }

    public int getFoodY() {
        return foodY;
    }

    private void checkForFood() {
        if (snake.getHead().getX() == foodX && snake.getHead().getY() == foodY) {
            spawnFood();
            snake.eat();
        }
    }

    private void checkBodyCollisions() {
        for (SnekPiece bodyPart : snake.getBody()) {
            if (bodyPart.getX() == snake.getHead().getX() && bodyPart.getY() == snake.getHead().getY()) {
                gameOver = true;
                return;
            }
        }
    }

    private void spawnFood() {
        boolean valid = true;
        do {
            valid = true;
            foodX = random.nextInt(SIZE - 1);
            foodY = random.nextInt(SIZE - 1);

            if (snake.getHead().getX() == foodX && snake.getHead().getY() == foodY) valid = false;

            for (SnekPiece piece : snake.getBody()) {
                if (piece.getX() == foodX && piece.getY() == foodY) valid = false;
            }
        } while (!valid);
    }
}